File: Mine/03Monopoly/MapImp.java
import java.util.*;

public class MapImp {
	protected Vector<Block> blocks;
	protected int row;
	protected int col;

	MapImp(int r, int c) {
		row = r;
		col = c;
		blocks = new Vector<Block>();
	}

	public void show() {
		Block[][] blockList = new Block[row][col];

		Block block = new WallBlock();
		for (int i = 0; i < blockList.length; i++){
			for (int j = 0; j < blockList[i].length; j++) {
				blockList[i][j] = block;
			}
		}

		for (Block specificBlock : blocks) {
			blockList[specificBlock.getRow()][specificBlock.getCol()] = specificBlock;
		}

		for (int i = 0; i < blockList.length; i++) {
			System.out.println();
            for (int j = 0; j < blockList[i].length; j++){
                System.out.print(blockList[i][j].toString());
            }
        }
	}

	public void addBlock(Block block) {
		blocks.add(block);
	}

	public Block getBlock(int index){
		return blocks.get(index);
	}

	public void setBlockVector(Vector<Block> blockVector) {
		blocks = blockVector;
	}

	public int getRow() {
		return row;
	}

	public int getCol() {
		return col;
	}
}

File: Teacher/05Richer/step2_src_完成12/mapimp.cpp
///======================================================================
///  Project:   Richer02
/// FileName:	mapimp.cpp
///     Desc:   Richer 01
///   Author:	Chen Wei
///======================================================================
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;
#include "mapimp.h"
#include "block.h"
#include "player.h"
MapImp::MapImp()
{
    //ctor
}

MapImp::~MapImp()
{
    //dtor
    for(Block* block : blocks) {
        delete block;
    }
}

void MapImp::addBlock(Block* block)
{
    blocks.push_back(block);
}
Block* MapImp::getBlock(int index)
{
    return blocks[index];
}
void MapImp::show() const
{
    //OutDevice & dev = OutDevice::getConsole();

    vector<Block*> temps(blocks);
    sort(temps.begin(), temps.end(), [](Block * b1, Block * b2) {
        if(b1->getRow() == b2->getRow()) {
            return b1->getCol() < b2->getCol();
        } else {
            return b1->getRow() < b2->getRow();
        }
    });

    /**
    CppPlayerIterator it(&Players::getPlayers());

    Block** bs = new Block*[it.count()];
    int i = 0;
    for(it.first();!it.isLast();it.next()) {
        Player * cur = it.current();
        bs[i++] = cur->getStand();
    }
    **/
    JavaPlayerIterator2 it(&Players::getPlayers());
    Block** bs = new Block*[it.count()];
    int i = 0;
    while(it.hasNext()) {
        Player* cur = it.next();
        bs[i++] = cur->getStand();
    }


    int row = 0;
    int col = 0;
    for(Block* block : temps) {
        int r = block->getRow();
        int c = block->getCol();
        for(int i = row; i < r; ++i) {
            cout << endl;
            col = 0;
        }
        for(int i = col; i < c; ++i) {
            cout << "   ";
        }
        if(block == bs[0]) {
            cout << "[" << block->name() << "]";
        } else if(block == bs[1]) {
            cout << "(" << block->name() << ")";
        } else {
            cout << " " << block->name() << " ";
        }
        row = r;
        col = c + 1;
    }
    cout << endl;
    delete[] bs;
}
